File: framework/src/play/src/main/scala/play/api/inject/guice/GuiceApplicationLoader.scala
/*
 * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
 */
package play.api.inject.guice

import com.google.inject.Module
import play.api.{ Application, ApplicationLoader, Configuration, Environment, OptionalSourceMapper }
import play.api.inject.{ bind, Injector => PlayInjector, Module => PlayModule }
import play.core.WebCommands

/**
 * An ApplicationLoader that uses Guice to bootstrap the application.
 */
class GuiceApplicationLoader(builder: GuiceApplicationBuilder) extends ApplicationLoader {

  // empty constructor needed for instantiating via reflection
  def this() = this(new GuiceApplicationBuilder)

  def load(context: ApplicationLoader.Context): Application = {
    builder
      .in(context.environment)
      .loadConfig(context.initialConfiguration)
      .bindings(modules(context: ApplicationLoader.Context): _*)
      .overrides(overrides(context): _*).build
  }

  def modules(context: ApplicationLoader.Context): Array[GuiceableModule] = {
    playModules(context: ApplicationLoader.Context).map(GuiceableModule.guiceable(_)) ++ guiceModules(context: ApplicationLoader.Context).map(GuiceableModule.guiceable(_))
  }

  /**
   * Override to provide your own Guice modules
   */
  def guiceModules(context: ApplicationLoader.Context): Array[Module] = {
    Array[Module]()
  }

  /**
   * Override to provide your own Play modules
   */
  def playModules(context: ApplicationLoader.Context): Array[PlayModule] = {
    Array[PlayModule]()
  }

  /**
   * Override to provide your own override modules
   */
  def overrides(context: ApplicationLoader.Context): Array[GuiceableModule] = {
    Array[GuiceableModule](
      bind[OptionalSourceMapper] to new OptionalSourceMapper(context.sourceMapper),
      bind[WebCommands] to context.webCommands)
  }

}

File: framework/src/play-guice/src/main/java/play/inject/guice/GuiceApplicationLoader.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.inject.guice;

import play.api.inject.guice.GuiceableModule;
import play.libs.Scala;
import play.Application;
import play.ApplicationLoader;

/**
 * An ApplicationLoader that uses Guice to bootstrap the application.
 *
 * Subclasses can override the <code>builder</code> and <code>overrides</code>
 * methods.
 */
public class GuiceApplicationLoader implements ApplicationLoader {

    /**
     * The initial builder to start construction from.
     */
    protected final GuiceApplicationBuilder initialBuilder;

    public GuiceApplicationLoader() {
        this(new GuiceApplicationBuilder());
    }

    public GuiceApplicationLoader(GuiceApplicationBuilder initialBuilder) {
        this.initialBuilder = initialBuilder;
    }

    @Override
    public final Application load(ApplicationLoader.Context context) {
        return builder(context).build();
    }

    /**
     * Construct a builder to use for loading the given context.
     *
     * @param context the context the returned builder will load
     * @return the builder
     */
    public GuiceApplicationBuilder builder(ApplicationLoader.Context context) {
        return initialBuilder
            .in(context.environment())
            .loadConfig(context.initialConfig())
            .overrides(overrides(context));
    }

    /**
     * Identify some bindings that should be used as overrides when loading an application using this context. The default
     * implementation of this method provides bindings that most applications should include.
     *
     * @param context the context that should be searched for overrides
     * @return the bindings that should be used to override
     */
    protected GuiceableModule[] overrides(ApplicationLoader.Context context) {
        scala.collection.Seq<GuiceableModule> seq = play.api.inject.guice.GuiceApplicationLoader$.MODULE$.defaultOverrides(context.asScala());
        return Scala.asArray(GuiceableModule.class, seq);
    }

}
